service/raft/raft_address_map.hh

```diff
@@ -112,83 +74,55 @@ class raft_address_map {
         // end.
         using list_type = bi::list<expiring_entry_ptr>;
 
-        explicit expiring_entry_ptr(list_type& l, timestamped_entry* e)
-            : _expiring_list(l), _ptr(e)
+        explicit expiring_entry_ptr(list_type& l, const raft::server_id& entry_id)
+            : _expiring_list(l), _last_accessed(Clock::now()), _entry_id(entry_id)
         {
-            _ptr->_last_accessed = Clock::now();
-            _ptr->set_lru_back_pointer(this);
+            _expiring_list.push_front(*this);
         }
 
         ~expiring_entry_ptr() {
-            if (lru_list_hook::is_linked()) {
-                _expiring_list.erase(_expiring_list.iterator_to(*this));
-            }
-            _ptr->_last_accessed = std::nullopt;
-            _ptr->set_lru_back_pointer(nullptr);
+            _expiring_list.erase(_expiring_list.iterator_to(*this));
         }
 
         // Update last access timestamp and move ourselves to the front of LRU list.
         void touch() {
-            _ptr->_last_accessed = Clock::now();
-            if (lru_list_hook::is_linked()) {
-                _expiring_list.erase(_expiring_list.iterator_to(*this));
-            }
+            _last_accessed = Clock::now();
+            _expiring_list.erase(_expiring_list.iterator_to(*this));
             _expiring_list.push_front(*this);
         }
         // Test whether the entry has expired or not given a base time point and
         // an expiration period (the time period since the last access lies within
         // the given expiration period time frame).
         bool expired(clock_duration expiry_period) const {
-            auto last_access_delta = Clock::now() - *_ptr->_last_accessed;
+            auto last_access_delta = Clock::now() - _last_accessed;
             return expiry_period < last_access_delta;
         }
 
-        timestamped_entry* timestamped_entry_ptr() {
-            return _ptr;
+        const raft::server_id& entry_id() {
+            return _entry_id;
         }
 
     private:
         list_type& _expiring_list;
-        timestamped_entry* _ptr;
+        clock_time_point _last_accessed;
+        const raft::server_id& _entry_id;
```

Comment:
We could store a `map_iterator` instead of `_entry_id`, won't need an additional lookup in `drop_expired_entries`

Reply:
Iterators are invalidated by inserting/erasing things to/from the map (if rehashing occurs)

Reply:
You are right, I confused it with a homemade hashmap from my prev job, it preserved iterators when rehashed.